plugins/node/opentelemetry-instrumentation-pulsar-client/src/proxies/consumer.ts

```diff
@@ -0,0 +1,211 @@
+import {Instrumentation} from '../instrumentation';
+import {SemanticAttributes} from "@opentelemetry/semantic-conventions";
+import {PulsarInstrumentationConfig} from "../types";
+
+const spanNames = {
+  beforeConsume: 'beforeConsume',
+  afterConsume: 'afterConsume',
+};
+
+type ConsumerListener = (
+  message: Pulsar.Message,
+  consumer: Pulsar.Consumer
+) => void | Promise<void>;
+
+export class ConsumerProxy implements Pulsar.Consumer {
+  private readonly _tracer: Tracer;
+  private readonly _instrumentationConfig: PulsarInstrumentationConfig;
+  private readonly _moduleVersion: string | undefined;
+  private readonly config: Pulsar.ConsumerConfig;
+  private consumer: Pulsar.Consumer;
+
+  private _lastSpan: Span | undefined;
+  private _lastAttributes: Attributes | undefined;
+
+  constructor(
+    _tracer: Tracer,
+    _instrumentationConfig: PulsarInstrumentationConfig,
+    _moduleVersion: string | undefined,
+    config: Pulsar.ConsumerConfig,
+    consumer: Pulsar.Consumer
+  ) {
+    this._tracer = _tracer;
+    this._instrumentationConfig = _instrumentationConfig;
+    this._moduleVersion = _moduleVersion;
+    this.config = config;
+    this.consumer = consumer;
+  }
+
+  async receive(timeout?: number): Promise<Pulsar.Message> {
+    this.closePreviousSpan();
+    const message = await this.consumer.receive(timeout);
+
+    // Postpone the span ending for the next time the user calls receive
+    this._lastSpan = extractSpanFromMessage(
+      this._tracer,
+      this._instrumentationConfig,
+      this._moduleVersion,
+      this.config,
+      message
+    );
+    this._lastAttributes = getAttributesFromMessage(message);
+    return message;
+  }
+
+  private closePreviousSpan() {
+    // User is done with the last message and called receive again.
+    if (this._lastSpan) {
+      this._lastSpan.end();
+      if (this._instrumentationConfig.trackBeforeAndAfterConsume) {
+        trackBeforeAndAfterConsume(this._tracer, spanNames.afterConsume, this._lastAttributes!);
+      }
+    }
+  }
+
+  acknowledge(message: Pulsar.Message): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.acknowledge(message);
+  }
+
+  acknowledgeId(messageId: Pulsar.MessageId): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.acknowledgeId(messageId);
+  }
+
+  negativeAcknowledge(message: Pulsar.Message): void {
+    this.closePreviousSpan();
+    this.consumer.negativeAcknowledge(message);
+  }
+
+  negativeAcknowledgeId(messageId: Pulsar.MessageId): void {
+    this.closePreviousSpan();
+    this.consumer.negativeAcknowledgeId(messageId);
+  }
+
+  acknowledgeCumulative(message: Pulsar.Message): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.acknowledgeCumulative(message);
+  }
+
+  acknowledgeCumulativeId(messageId: Pulsar.MessageId): Promise<null> {
+    this.closePreviousSpan();
+    return this.consumer.acknowledgeCumulativeId(messageId);
+  }
```

Comment:
There is some useful information here, like if the acknowledge is positive or negative (which indicates an error in processing the message that the operator might find useful). There is also the messageId which is not recorded and I guess can be handy for debugging and root causes investigations. The acknowledge is possibly cumulative or for a single message.

This can also be addressed by later PRs, just drawing attention to a possible improvement for usability.